lib/scala/filewatcher/src/main/java/org/enso/filewatcher/JWatcher.java

```diff
@@ -0,0 +1,180 @@
+    this.root = root;
+    this.eventCallback = eventCallback;
+    this.exceptionCallback = exceptionCallback;
+    this.watchService = watchService;
+  }
+
+  public void start(Executor executor) {
+    if (closed) {
+      throw new IllegalStateException("Watcher has already been stopped.");
+    }
+    System.out.println("[JWatcher] Starting watcher for root " + root.toAbsolutePath());
+    try {
+      var watchKey =
+          root.register(
+              watchService,
+              StandardWatchEventKinds.ENTRY_MODIFY,
+              StandardWatchEventKinds.ENTRY_CREATE,
+              StandardWatchEventKinds.ENTRY_DELETE,
+              StandardWatchEventKinds.OVERFLOW);
+      watchedDirs.put(root, watchKey);
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+    executor.execute(this::eventLoop);
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+    try {
+      watchService.close();
+    } catch (IOException e) {
+      exceptionCallback.accept(new JWatcherError(e));
+    }
+    for (var watchKey : watchedDirs.values()) {
+      watchKey.cancel();
+    }
+  }
+
+  private void eventLoop() {
+    try {
+      while (!closed) {
+        var iterator = watchedDirs.entrySet().iterator();
+        while (iterator.hasNext()) {
+          var entry = iterator.next();
+          var dir = entry.getKey();
+          var watchKey = entry.getValue();
+          for (var event : watchKey.pollEvents()) {
+            dispatchEvent(event, dir);
+          }
+          var valid = watchKey.reset();
+          if (!valid) {
+            iterator.remove();
+          }
+        }
+      }
+    } catch (Throwable e) {
+      var err = new JWatcherError(e);
+      exceptionCallback.accept(err);
+    }
+  }
+
+  /**
+   * @param event
+   * @param dir The directory where the event occurred. This directory must have already been
+   *     registered by the watchservice.
+   */
+  private void dispatchEvent(WatchEvent<?> event, Path dir) {
+    var eventPath = (Path) event.context();
+    var absolutePath = dir.resolve(eventPath);
+    var isDir = Files.isDirectory(absolutePath);
+    var isRepeated = event.count() > 1;
+    var eventType = deduceType(event);
+    System.out.println(
+        "[JWatcher] Received event, kind:"
+            + event.kind()
+            + ", path: "
+            + absolutePath
+            + ", isDir: "
+            + isDir
+            + ", isRepeated: "
+            + isRepeated);
+    if (isRepeated) {
+      return;
+    }
+    if (eventType == EventType.CREATE && isDir) {
+      registerWatchService(absolutePath);
+    } else if (eventType == EventType.DELETE && isDir) {
+      cancelWatch(absolutePath);
+    }
+    if (eventType != null) {
+      var convertedEvent = new JWatcherEvent(absolutePath, eventType);
+      eventCallback.accept(convertedEvent);
+    }
+  }
+
+  private EventType deduceType(WatchEvent<?> event) {
+    return switch (event.kind().name()) {
```

Comment:
- It is a bit strange to `switch` on the `name()`. 
- Isn't it better to just cascade few `if` statements and `if (event.kind() == StandardWatchEventKinds.ENTRY_CREATE`?
- PS: Originally I wanted to suggest:
```suggestion
    return switch (event.kind()) {
```
- but now I know that `kind()` isn't an enum

Reply:
Refactored in c87dd0682cc9b9ca608bf83a5ebef96b9036411c